[content-visibility] Improve performance of grid based perf tests

rwlbuis opened the pull request:
#### 324b1bf4b75f0f8916f982084ef49cfb34a957fb
<pre>
[content-visibility] Improve performance of grid based perf tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=261825">https://bugs.webkit.org/show_bug.cgi?id=261825</a>

Reviewed by NOBODY (OOPS!).

Improve performance of grid based perf tests by skipping parts of the track sizing algorithm.
As an example, on my mbp grid-item-width-changes.html results change from 738ms to 16ms.

To verify that the optimization does not break getComputedStyle on grid items content-visibility-100.html is added.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-100-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-100.html: Added.
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::run):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/324b1bf4b75f0f8916f982084ef49cfb34a957fb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40912 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19925 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43290 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43471 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37003 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43219 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22931 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17256 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33905 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41486 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16865 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35263 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14528 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14634 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36254 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44766 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37119 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36564 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40314 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15727 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12911 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38683 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17346 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/35534 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17397 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16990 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
Question for this patch in general: Is this correct when ContentVisibilityForceLayout is on ? Or is this unreachable/irrelevant code when ContentVisibilityForceLayout is on?

alanbaradlay commented on Source/WebCore/rendering/RenderElement.h:
```diff
@@ -294,6 +294,7 @@ class RenderElement : public RenderObject {
     bool createsNewFormattingContext() const;
 
     bool isSkippedContentRoot() const;
+    bool isSkippedContentRootForLayout() const;
```
does this mean we have to call isSkippedContentRootForLayout() when inside layout instead of calling isSkippedContentRoot()? if so, how to tell whether you are _inside_ layout?

rwlbuis replied:
Sorry, I got sidetracked with other tasks. Yes, isSkippedContentRootForLayout() is the safe variant to call when in layout. Usually the places to call this from are pretty clear, layoutBlock and layoutPositionedObject for example. Do you foresee call sites where it is not clear whether we are in layout or not? Regarding the grid changes, the expectation is that the track sizing algorithm is only called when laying out.

alanbaradlay commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1708,6 +1711,9 @@ void GridTrackSizingAlgorithm::run()
 
+    if (m_renderGrid->isSkippedContentRoot())
+        return;
+
```
do we need to run step #1 and #2 for skipped content?

rwlbuis replied:
It used to be needed in the past, but now that I try locally I see no problem moving the check before Step 1. Let's see what the buildbots say.

rwlbuis committed:
[content-visibility] Improve performance of grid based perf tests
https://bugs.webkit.org/show_bug.cgi?id=261825

Reviewed by NOBODY (OOPS!).

Improve performance of grid based perf tests by skipping parts of the track sizing algorithm.
As an example, on my mbp grid-item-width-changes.html results change from 738ms to 16ms.

To verify that the optimization does not break getComputedStyle on grid items content-visibility-100.html is added.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-100-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-100.html: Added.
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::run):
